### DISCUSSION TITLE
Multi DB support in 1.8?

### DISCUSSION THREAD

**Discussion by friarpat** [medusajs/medusa#3819] [Q&A]
Prior to 1.8, a middleware could change which database was to be accessed by the various services by doing something like:

        req.scope.register({manager: asValue(manager)});
        req.scope.cache.clear();
        req.scope = req.scope.createScope() as any;

...passing in the manager that was connected to the desired database.
As of 1.8, this functionality no longer seems to work?   I can perform the above steps.  I can verify in a "later" middleware that the manager in the DI is indeed pointing to the desired database.  I can successfully query for data from that database using the manager from the DI.

...but getting a list of products via store/products always returns data from the default database regardless what value is in the req.scope.

Is there a new way of handling this in 1.8?

Thanks!

**Comment by srindom** [comment]
Maybe @adrien2p has some inputs on this? :)

@friarpat - out of curiosity, what do you use this "dynamic db control" for?

**Reply by adrien2p** [reply]
This is one of the multi tenancy approach @srindom , if you want one db per tenant.

@friarpat could you share a bit more code, where did you apply the middleware, how, plua your entire middleware etc so that i can get a better idea.

also, now with the new typeorm version the repositories are built from the datasource directly so you might need a bit more work in order to attach the repositories the the different datasources

**Reply by friarpat** [reply]
I created a Singleton service under src/services to capture the multiple DataSource connections:
```
class DataSourceManagerService extends TransactionBaseService {
    static LIFE_TIME = Lifetime.SINGLETON;
    connections: Map<string, DataSource> = new Map();

  public async getOrCreateConnection(defaultManager: EntityManager, req: Request){
    // extract the tenant/vendor from the req and then return either the existing DataSource or
    // or create the new DataSource...all stuff you've seen before!!!
        let connection!: DataSource;
        if (this.connections.has(tenant.code)){
            console.log("Connection was already in the connections map");
            connection = await this.connections.get(tenant.code);
        }
        else{
            const db_entities = req.scope.resolve('db_entities') as any [];
            connection = new DataSource({
                name: tenant.code,
                type: tenant.database_config.database_type as any,
                url: tenant.database_config.database_url,
                database: tenant.database_config.database_database,
                extra: tenant.database_config.database_extra || {},
                entities: db_entities,
                logging: configModule.projectConfig.database_logging ?? false,
            });

            this.connections.set(tenant.code, connection);
        }
 
       return await new Promise((resolve) => {
            connection.isInitialized
                ? resolve(connection.manager)
                : connection.initialize().then((conn) => resolve(conn.manager));
        });

```

That service is called by a middleware created under src/api/middlewares.  

```
   // Created a class ...does NOT extend the TransactionBaseService
    export class TenantService{
  // ....
   public async multitenancyMiddleware(req: Request, res: Response, next: NextFunction): Promise<void> {
        const defaultManager: EntityManager = req.scope.resolve("manager");

       // grab the above service (DataSourceManagerService ) out of the DI
        const datasourceMgr = req.scope.resolve("datasourceManagerService");

      // call the service to get the data source mgr
      let manager: EntityManager;
       try {
            manager = await datasourceMgr.getOrCreateConnection(defaultManager, req);
          } catch (e) {
            return next(e);
        }

    // register the returned manager, clear the cache, create a new scope 
    //...which "should" generate all of the services again with the new datasource manager.
    req.scope.register({manager: asValue(manager)});
    req.scope.cache.clear();
    req.scope = req.scope.createScope() as any;

```
in the src/api/middlewares/index.ts, added the following:

```
  const mm = new TenantService();
  router.use("/", mm.multitenancyMiddleware);

```

Using postman, to do a GET localhost:9000/store/products

- Logging shows I'm setting up the datasource to the correct db.
- Once the scope lines are ran in the middleware, I can pull the manager back out of the DI and am able to pull the products from the appropriate database.
- I set up an additional router.use in the middlewares index.ts. When it runs, it has the correct datasource from the DI pointing to the expected database.
- ...but the results to the postman query are always from the default database.
- Modifying medusa-config.js to point to one of the "other" databases allows me to get back data for that vendor - but only that vendor.  It's now the new default database.

I'm a little stumped in that I would have expected the req.scope = req.scope.createScope() to generate a new set of service classes - but using the replaced manager.

Thanks for the help.

**Reply by adrien2p** [reply]
ahah alright you just have taken the extender code

What I am saying is that now the repositories are using the globally exported dataSource from the database loader when they are created and then the manager resolve the repository. This is why I am saying that it might required to do something extra for the repository now that they are linked to the dataSource. You might need to open an issue in the typeorm repository in order to ask how to change the repository datasource on the fly in a multi tenancy architecture involving multiple db's.

You might need something like resolving the repositories from the container, and update their datasource, i have not tested but it might be something. But it might be better to ask typeorm directly as there is not much information on their side about that topic

**Reply by friarpat** [reply]
Oof!   Don't know why I didn't see it before, but in the Medusa dependency documentation I see:

Core services by default have the SINGLETON lifetime. However, some have a different lifetime which is indicated in this table. Custom services, including services in plugins, by default have the SCOPED lifetime, unless defined differently within the custom service.

I'm assuming that's why I'm not getting a new instance of the services when the scope is recreated for the request?
Is the solution to extend all core services and make the lifetime SCOPED?

**Reply by friarpat** [reply]
Sorry - didn't see your response.   I'll take this up with typeorm.

**Reply by adrien2p** [reply]
no this is not why, I ve used cache clear in order to force the entire container to be re resolved exactly for the singleton issue. So that is why it was working prior medusa 1.8

**Reply by friarpat** [reply]
uhhh-k.   This is a "design out loud" post...but seeing if I'm on the right track or going to cause someone on the team to faint!  

Assumption 1:  All of those repositories in the DI get recreated on a createScope() call.
Assumption 2:  The repositories are using the database loader to get the datasource for the repository when the new scope is created. 
Assumption 3:  I can override a loader in Medusa.

Provided the above assumptions are valid - grab your air sickness bag...

- Instead of having the database loader ONLY load configuration from  medusa-config.js each time it's called, have it load it from the DI as "ActiveDBConfig". ...if that's not present, then load from medusa-config.js
- default config is from the medusa-config
- Set another DI value with "ActiveDBName"
- default to "default"
- Create a DI value that contains a map of DataSourceConnectors.   
- Extend the database loader to function similarly to the DataSourceManagerService indicated above.   That is, if the "ActiveDBName" is in the DataSourceConnectors - return the datasource.  If not, use the "ActiveDBConfig" to create the DataSourceConnector, stash it in the DI's DataSourceConnectors map, then return the datasource.
- The middleware would now register the ActiveDBConfig in case this is a new connection, set the ActiveDBName in the request's copy of the scope, flush the cache, then create the new scope.  
- The extended database loader should now either create the new connection and cache it, or, simply return the cached dataSource from the DI as the various repositories are created/instantiated as part of the createScope() call.

Provided you survived the above description - which DI scope would the database loader be using to look for cached dataSources when doing the createScope() call?   Wouldn't want to change the "base" DI scoped ActiveDBConfig and ActiveDBName, but would want to share all of the cached dataSources across all "users".   Is it enough to make the DataSourceConnectors a SINGLETON, and the other two SCOPED?

**Reply by adrien2p** [reply]
I think that in fact you would only have to resolve all the repositories and change the manager and queryRunner by assigning the new one from the new datasource. But there might be some risk and i would advise you to wait for typeorm answer or look at typeorm implementation if they don’t come back to you, or use a different multi tenancy approach such as multi schema for example

**Comment by CervarlCG** [comment]
Hello, did you find a solution for this?

**Comment by friarpat** [comment]
I did not.  I attempted several things…but in the end – at that time, it seemed like Medusa moved away from supporting anything “multi” with their latest version.  Our timelines were too short to attempt to sway typeorm for Medusa’s needs.  All of this may have changed by now?  We had to abandon Medusa and go with a solution that was less desirable.  Too bad, as I think there are lots of folks out there that need something like this!

**Reply by octalpixel** [reply]
@friarpat what solution have you considered now ?

**Comment by vkislichenko** [comment]
I'm having similar problems, and I think implementation of an option to "properly" scope modules - would bring solution a bit closer.
I've opened [dedicated](https://github.com/medusajs/medusa/discussions/4546) feature request for it.

for now my plan is to force "scoped" lifetime to all registrations in main container + override (similarly how medusa-extender did it) modules loader to implement "native" cradle usage

**Comment by amirping** [comment]
I'm willing to find a solution to implement the multi-tenancy design with Medusa, but yeah looks very hard now. it was easier with the extender but it's not here anymore.

**Comment by voronovmaksim** [comment]
Are there any ways to do this?
